lazy feature flag evaluation

thenamankumar opened the pull request:
Closes https://github.com/sourcegraph/sourcegraph/issues/35541.
Closes https://github.com/sourcegraph/sourcegraph/issues/35542.
Closes https://github.com/sourcegraph/sourcegraph/issues/35543.
Closes https://github.com/sourcegraph/sourcegraph/issues/35544.
## Test plan

This PR brings lazy feature flag evaluation. Currently, all feature flags are evaluated the first time user lands on the page and not when the user actually participates in the experiment. 

In the old API, there is no way to evaluate a feature flag on-demand in a lazy manner. To support the purpose of this PR, 
- We removed the old method FromContext which combined both evaluating all feature flags and returning the values at once 
- Created two separate functions:
  -  `EvaluateForActorFromContext` which on-demand evaluates a single feature flag and returns its value. It is used by frontend/backend clients to evaluate flags and perform actions accordingly.
  -  `GetEvaluatedFlagsFromContext` returns the list of already evaluated feature flags for the user. It is only used to pass the feature flags context to all the event logs we record.

This PR implements a new graphql field `evaluateFeatureFlag` which evaluates a feature flag and returns the value using `EvaluateForActorFromContext`. It also stores the already evaluated feature flags in Redis cache and passes the evaluated feature flags' context to all the event logs using `GetEvaluatedFlagsFromContext`.

### Why store evaluated feature flags in Redis?
Feature flags are evaluated not just for logged in users but un-authenticated visitors as well. That means we need to store feature flag value for each of those monthly visitors separately for each feature flag, delete them if/when the flag is deleted and also query them when logging events. Redis provides easy management of this cache when this data is only used to pass onto event logs and is not used in any business logic.

keegancsmith reviewed (commented):
Thanks for what looks like it fixes a lot of open issues. Nice work! I just have some very quick drive-by comments:
- why are you introducing a redis cache? Have we noticed overhead via speaking to the DB?
- the api change from `featureflag.FromContext` to `featureflag.GetEvaluatedFlagsFromContext` doesn't read very nicely. Do we need such a long function name?
- why do we now have a `EvaluateForActorFromContext` function vs the old API?
- for such a big PR, there isn't much motivation / decisions described in the description.

thenamankumar commented:
> Thanks for what looks like it fixes a lot of open issues. Nice work! I just have some very quick drive-by comments:
> 
>     * why are you introducing a redis cache? Have we noticed overhead via speaking to the DB?
> 
>     * the api change from `featureflag.FromContext` to `featureflag.GetEvaluatedFlagsFromContext` doesn't read very nicely. Do we need such a long function name?
> 
>     * why do we now have a `EvaluateForActorFromContext` function vs the old API?
> 
>     * for such a big PR, there isn't much motivation / decisions described in the description.

@keegancsmith 

1. Feature flags are evaluated not just for logged in users but un-authenticated visitors as well. That means we need to store feature flag value for each of those monthly visitors separatly for each feature flag, delete them if/when the flag is deleted and also query them when logging events. Redis gives an easy management of this cache when this data is only used to pass onto event logs and is not used in any business logic.
2. As there is a fundamental difference between `FromContext` and the new API `GetEvaludatedFlagsFromContext`, i.e. the new API only returns the already evaluated feature flags as compared to the previous one which evaluates all the feature flags and then returns the value for each of them. To indicate the change, we tend to not go with the old name. How about just naming it `EvaluatedFromContext`?
3. About `EvaluateForActorFromContext`. In the old API, there is no way to evaluate a feature flag on-demand in a lazy manner. To support the purpose of this PR, we removed the old method `FromContext` which combined both evaluating all feature flags and returning the values at once, and created two separate functions. First `EvaluateForActorFromContext` which on-demand evaluates a single feature flag and returns its value. Second `GetEvaluatedFlagsFromContext` which returns the list of already evaluated feature flags for the user. `EvaluateForActorFromContext` is used by frontend/backend clients to evaluate flags and perform actions accordingly while `GetEvaluatedFlagsFromContext` is only used to pass the feature flags context to all the event logs we record.

keegancsmith commented on internal/featureflag/cache.go:
```diff
@@ -0,0 +1,72 @@
+)
+
+var (
+	pool = redispool.Store
```
you are using `redispool.Store`, which will not expire entries. From your description of your use of redis, you want a cache which will expire entries? Or is there code somewhere to delete items?

thenamankumar replied:
There is a function `ClearFlagFromCache` defined in the same file which is used when a feature flag is deleted to clear the cache in redis here: https://github.com/sourcegraph/sourcegraph/pull/35359/files#diff-341cacde31fa80b70a9ca48613f9f83f5f117b9c82bfde6476632bcb1eda2fc3R162

keegancsmith commented on internal/featureflag/middleware.go:
```diff
@@ -32,62 +35,47 @@ func Middleware(ffs Store, next http.Handler) http.Handler {
+}
 
-	return FlagSet(make(map[string]bool))
+func EvaluateForActorFromContext(ctx context.Context, flagName string) (result bool) {
```
This function deserves a docstring. In particular you should call out that if the context doesn't have WithFlags set we silently return nil. In the old imlpementation we would panic (which isn't great, but atleast was noticeable)

thenamankumar replied:
Added docstring

keegancsmith reviewed (commented):
Thanks for updating the PR description and responding :)

Another change I noticed is previously we had a way to set a default value at each call site if the feature flag was unset. That seems to be missing? How do we now code in when changing a feature flag to default true?

> As there is a fundamental difference between `FromContext` and the new API `GetEvaludatedFlagsFromContext`...

You have updated all the callsites, so the old FromContext doesn't exist. Personally I'd prefer the more ergonomic name, since this will be the most commonly used API for feature flags in the backend. A common name in go when a package is doing this sort of thing is just [`FromContext`](https://sourcegraph.com/search?q=context:%40keegan+func+%5CbFromContext%5Cb&patternType=regexp).

> How about just naming it EvaluatedFromContext?

That is better, since it also avoids stutter (sayings Flags twice at call sites, eg `flags.$$$Flags$$$`. But, as above I'd argue with going with the common idiom of just FromContext. I'm unsure what benefit saying "Evaluated" brings readers of call sites.

> 3. About `EvaluateForActorFromContext`. In the old API, there is no way to evaluate a feature flag on-demand in a lazy manner...

Ok great that was the context I was missing. Thank you :)

> Why store evaluated feature flags in Redis

I think you should make clear in your implementation (maybe a docstring) is that we don't read redis for feature flags. The use of redis is purely for the stats/reports you mentioned.

cc @camdencheek

thenamankumar commented:
@keegancsmith I have renamed `GetEvaludatedFlagsFromContext` to `FromContext` and added necessary docstring as well. Can you please take another look at the code?

camdencheek commented on internal/featureflag/cache.go:
```diff
@@ -0,0 +1,72 @@
+
+func getVisitorIDForActor(a *actor.Actor) (string, error) {
+	if a.IsAuthenticated() {
+		return fmt.Sprintf("uid_%v", a.UID), nil
```
nit: it's preferable to use formatting args specific to the type so the format is more predictable and changes will be caught by static analysis. 

```suggestion
		return fmt.Sprintf("uid_%d", a.UID), nil
```

thenamankumar replied:
resolved

camdencheek commented on internal/featureflag/cache.go:
```diff
@@ -0,0 +1,72 @@
+	} else if a.AnonymousUID != "" {
+		return "auid_" + a.AnonymousUID, nil
+	} else {
+		return "", errors.New("UID/AnonymousUID are emptry for the given actor.")
```
return "", errors.New("UID/AnonymousUID are empty for the given actor.")

thenamankumar replied:
resolved

camdencheek reviewed (commented):
I'm still a bit confused why we're using Redis here, especially since we're apparently using it as a persistent DB rather than as a cache. Wouldn't it make more sense to store the set of evaluated feature flags in Postgres? That way we have a more explicit data model for the evaluated flag values and we can take advantage of foreign key relations for cleanup. Additionally, I'm concerned about adding yet another infrastructure dependency here since feature flags need to be usable across all our services (e.g., I have no idea whether `gitserver` has access to redis, but I know it has access to the db).

camdencheek commented on internal/featureflag/middleware.go:
```diff
@@ -32,62 +35,55 @@ func Middleware(ffs Store, next http.Handler) http.Handler {
+// set the evaluated flags to redis cache which later can be used to pass
+// feature flags context to event logs.
+func EvaluateForActorFromContext(ctx context.Context, flagName string) (result bool) {
+	result = false
```
Small point of feedback: defaulting to `false` without providing the user any way to know that the default was hit is not ideal. A frequent pattern I use is:
1) Implement a feature behind a feature flag, enabling it only for myself
2) Enable the feature for all users on sourcegraph.com, monitor the rollout
3) Update the feature flag calls to default `true` (`GetBoolOr("my-flag", true)`)
4) Wait for a version cut/release
5) Remove the old code

This allows me to set a new feature to default to `true` for customers while still providing customers an escape hatch to disable it if something goes wrong on their instance (by creating the feature flag and setting it to `false`). 

By defaulting to `false` here, this pattern is not possible because a feature flag that doesn't exist will always evaluate to `false`, and the caller doesn't know that it doesn't exist.

thenamankumar replied:
@camdencheek right now the feature flagging system's interface does not provide nullable feature flag values but only boolean. We have already planned to work on that so that we can return `nil`, `false`, `true` to the client. And along with that, we will also add `defaultValue` args to the function.

This will be handled separately.

camdencheek commented on internal/featureflag/middleware.go:
```diff
@@ -32,62 +35,55 @@ func Middleware(ffs Store, next http.Handler) http.Handler {
-		flags, err := f.ffs.GetUserFlags(ctx, a.UID)
-		if err == nil {
-			return FlagSet(flags)
+		if flag, err = f.ffs.GetUserFlag(ctx, a.UID, flagName); flag != nil {
+			setEvaluatedFlagToCache(flagName, a, *flag)
 		}
```
> having lazy feature flag evaluation is a business requirement

Performance is also a business requirement 🙂 

> feature flags evaluating on the same page [...] we can/are going to improve this by implementing a data-loader

Reminder that feature flags are also used fairly extensively in the backend, and using them for A/B tests are not their only use case. A data loader is not an option in the backend.

<hr/> 

I'm quite concerned about the separate db request per feature flag evaluation. This significantly limits how we can use them in the backend which is pretty sensitive to adding a bunch of db roundtrips in the critical path.

However, I don't think this is actually necessary to fulfil your requirements. The act of evaluating feature flags is stateless: we don't store what feature flags are assigned to a user in the db, we only store overrides.

Because of this, I think a model that fetches all feature flags on the first request (how it was before) is still best because we then we have all the feature flags available to query from. Then, whenver a _specific_ feature flag is requested (the new `EvaluateForActorFromContext`), we can store the evaluated result in Redis so that only the feature flags that have actually been evaluated for a user are saved in the event log. As you said, Redis operations are pretty cheap, so this is a much lower price to pay. 

The performance optimizations that were in place existed for a reason -- calling to the db on every feature flag was adding an unreasonable amount of latency.

In any case, would you please instrument feature flag evaluation with tracing so that we can see its impact?

erzhtor replied:
>Because of this, I think a model that fetches all feature flags on the first request (how it was before) is still best because we then we have all the feature flags available to query from. Then, whenver a specific feature flag is requested (the new EvaluateForActorFromContext), we can store the evaluated result in Redis so that only the feature flags that have actually been evaluated for a user are saved in the event log. As you said, Redis operations are pretty cheap, so this is a much lower price to pay.

Hey, @camdencheek! Thanks for the great suggestion, like it. I'm currently working on implement it this way. Will let you know once finishing it.

camdencheek replied:
Great! Thank you! Let me know if you need anything from me. I should be a bit more available this week 😄

erzhtor reviewed (commented):
@camdencheek, rewrote based on your suggestion:
1. Backend feature flag API is unchanged (`FromContex(ctx).GetBool[Or]`), and all flags are evaluated all at once as previously
1. Evaluated flags are set to Redis when `GetBool/GetBoolOr` is called
    - Changed the `FlagSet` type to store `*actor.Actor`, which is required to set the flag in Redis
    - Also, we store only flags that exist in database
    - The feature flag is removed from Redis when it is deleted 
1. Added a separate `EvaluatedFlagSet` type and `GetEvaluatedFlagSet` type specifically for logging usage purposes
1. Added default value support to frontend API as well, similar to backend

I would appreciate it if you could take/feedback on the implementation overall.

P.S: Also, would be grateful if you could help/point out to tests which mocked Redis. I wanted to add additional tests to check that flags are correctly set/removed to/from Redis, although manually tested. I tried to use similar to [this](https://github.com/sourcegraph/sourcegraph/blob/7148009913d77239bb24e79274682cc1e0866c32/cmd/frontend/internal/usagestatsdeprecated/usage_stats_test.go#L373) approach but doesn't seem to work.

camdencheek commented on internal/featureflag/middleware_test.go:
```diff
@@ -45,27 +47,82 @@ func TestContextFlags(t *testing.T) {
+
+	// Make sure user1 flags are set
+	flags := FromContext(ctx)
+	v, ok := flags.GetBool("user1")
```
I made a [draft PR](https://github.com/sourcegraph/sourcegraph/pull/36679) with an example of using `go-mockgen` to create a mock redis connection for testing. Feel free to merge or just copy stuff from that PR, I just thought it was the easiest way to give an example of how mocking out redis was possible. 

In an ideal world, we wouldn't be using a global redis pool (we'd hold a reference to a redis pool on `FlagSet`), but all our redis stuff is global right now, so I think this is probably fine.

camdencheek replied:
Alternatively, it looks like [this library](https://pkg.go.dev/github.com/rafaeljusto/redigomock) does pretty much what I wrote a hacky version of in that PR.

erzhtor replied:
Thank you, used partially your draft PR + library you shared to mock Redis! 👍

camdencheek commented on internal/featureflag/flagset.go:
```diff
@@ -3,25 +3,51 @@ package featureflag
 )
 
-type FlagSet map[string]bool
+type EvaluatedFlagSet map[string]bool
```
Could we add some comments to these types to explain what they are and why they are different?

camdencheek commented on internal/featureflag/cache.go:
```diff
@@ -0,0 +1,73 @@
+		return
+	}
+
+	c.Do("HSET", getFlagCacheKey(flagName), visitorID, fmt.Sprintf("%v", value))
```
We can be a bit more precise about the fact that we require a bool here:
```suggestion
	c.Do("HSET", getFlagCacheKey(flagName), visitorID, strconv.FormatBool(value))
```

camdencheek commented on internal/featureflag/flagset.go:
```diff
@@ -3,25 +3,51 @@ package featureflag
+func (f EvaluatedFlagSet) String() string {
+	var sb strings.Builder
+	for k, v := range f {
+		if v {
```
nit: is this supposed to skip all feature flags that evaluated to `false`?

erzhtor replied:
Not sure, to be honest. I've reused the existing FlagSet.String() implementation. It looks like it was added for tracing purposes in https://github.com/sourcegraph/sourcegraph/pull/35609 PR recently.

camdencheek reviewed (approved):
Pending remaining comments and redis tests (see [draft PR](https://github.com/sourcegraph/sourcegraph/pull/36679/files)), this looks good to me 👍

thenamankumar commented on client/web/src/featureFlags/lib/FeatureFlagClient.ts:
```diff
@@ -1,82 +1,89 @@
+                map(value => {
+                    // Use local feature flag override if exists
+                    const overriddenValue = getFeatureFlagOverride(flagName)
+                    if (overriddenValue === null) {
+                        return value
+                    }
```
not needed when already doing the blow checks
